telegram/ext/filters.py

```diff
@@ -372,1895 +363,1807 @@ def name(self, name: str) -> NoReturn:
-        ``MessageHandler(Filters.dice.football([5, 6]), callback_method)``.
+        ``MessageHandler(filters.Dice.Football([5, 6]), callback_method)``.
 
     Note:
         Dice messages don't have text. If you want to filter either text or dice messages, use
-        ``Filters.text | Filters.dice``.
+        ``filters.TEXT | filters.DICE``.
 
     Args:
-        update (:obj:`int` | Tuple[:obj:`int`] | List[:obj:`int`], optional):
-            Which values to allow. If not specified, will allow any dice message.
+        values (:obj:`int` | Tuple[:obj:`int`] | List[:obj:`int`], optional):
+            Which values to allow. If not specified, will allow the specified dice message.
 
     Attributes:
-        dice: Dice messages with the emoji 🎲. Passing a list of integers is supported just as for
-            :attr:`Filters.dice`.
-        darts: Dice messages with the emoji 🎯. Passing a list of integers is supported just as for
-            :attr:`Filters.dice`.
-        basketball: Dice messages with the emoji 🏀. Passing a list of integers is supported just
-            as for :attr:`Filters.dice`.
-        football: Dice messages with the emoji ⚽. Passing a list of integers is supported just
-            as for :attr:`Filters.dice`.
-        slot_machine: Dice messages with the emoji 🎰. Passing a list of integers is supported just
-            as for :attr:`Filters.dice`.
-        bowling: Dice messages with the emoji 🎳. Passing a list of integers is supported just
-            as for :attr:`Filters.dice`.
-
-            .. versionadded:: 13.4
-
+        DICE: Dice messages with the emoji 🎲. Matches any dice value.
+        Dice: Dice messages with the emoji 🎲. Supports passing a list of integers.
+        DARTS: Dice messages with the emoji 🎯. Matches any dice value.
+        Darts: Dice messages with the emoji 🎯. Supports passing a list of integers.
+        BASKETBALL: Dice messages with the emoji 🏀. Matches any dice value.
+        Basketball: Dice messages with the emoji 🏀. Supports passing a list of integers.
+        FOOTBALL: Dice messages with the emoji ⚽. Matches any dice value.
+        Football: Dice messages with the emoji ⚽. Supports passing a list of integers.
+        SLOT_MACHINE: Dice messages with the emoji 🎰. Matches any dice value.
+        SlotMachine: Dice messages with the emoji 🎰. Supports passing a list of integers.
+        BOWLING: Dice messages with the emoji 🎳. Matches any dice value.
+        Bowling: Dice messages with the emoji 🎳. Supports passing a list of integers.
     """
 
-    class language(MessageFilter):
-        """Filters messages to only allow those which are from users with a certain language code.
+    __slots__ = ()
+    # Partials so its easier for users to pass dice values without worrying about anything else.
+    DICE = _DiceEmoji(emoji=DE.DICE)  # skipcq: PTC-W0052
+    Dice = partial(_DiceEmoji, emoji=DE.DICE)  # skipcq: PTC-W0052
```

Comment:
How well does does this play with type checking and auto completion? e.g. what does mypy think the type of `Dice.Dice` is? `reveal_type(filters.Dice.Dice)` should tell you if you run mypy on an example file. does mypy accept passing `filters.Dice.Dice` as first argument to `MessageHandler`, i.e. where a `BaseFilter` is expected?

Reply:
okay not sure what's happening but when I run mypy with `reveal_type(filters.Dice.Dice)` or anything from filters it says:
`error: Module has no attribute "Dice"`

Reply:
`foo.py`:

```python
from telegram.ext import filters, MessageHandler


print(filters.Dice())
reveal_type(filters.Dice())
print(filters.Dice.Dice)
reveal_type(filters.Dice.Dice)
print(filters.DICE)
reveal_type(filters.DICE)

MessageHandler(filters.Dice.Dice, callback=lambda u, c: 1)
```

running `mypy foo.py` gives

```python
foo.py:5: note: Revealed type is "telegram.ext.filters.Dice"
foo.py:7: note: Revealed type is "functools.partial[telegram.ext.filters._DiceEmoji]"
foo.py:9: note: Revealed type is "telegram.ext.filters.Dice"
foo.py:12: error: Cannot infer type argument 1 of "MessageHandler"  [misc]
foo.py:12: error: Argument 1 to "MessageHandler" has incompatible type "partial[_DiceEmoji]"; expected "BaseFilter"  [arg-type]
```

Also auto-complete for `filters.Dice.Dice` also doesn't work …

![image](https://user-images.githubusercontent.com/22366557/139929477-64a05fa4-7e96-486d-bbbe-d857f6a641ef.png)

It's a shame, that would have been really nice!

Reply:
😞 found the issues - https://youtrack.jetbrains.com/issue/PY-25698, https://github.com/python/mypy/issues/1484.

Sucks that type checkers and autocompletion weigh us down sometimes..

Guess I'll make these into nested classes then..

Reply:
mh, I had an error in the example: the last line should be something like

```python
MessageHandler(filters.Dice.Dice(values=[7]), callback=lambda u, c: 1)
```

which changes the error to 

```python
foo.py:12: error: Cannot infer type argument 1 of "MessageHandler"  [misc]
```
:D 

I tried something with a class-factory, but annotation the return value is the hard part, then …

> Sucks that type checkers and autocompletion weigh us down sometimes..

Depends on the POV - sure a bit more code for now but after all we want the lib to be easy to use :)